File: lib/prompts/timeline.ts
import type { MedicalEntities, TimelineEntry } from "@/lib/events";

export const TIMELINE_SYSTEM = `You are a clinical timeline agent. Produce a concise, chronological list of key clinical events from the encounter.

Return ONLY a raw JSON array (no markdown):
[{ "id": string, "timestamp": string (ISO 8601), "summary": string, "source": "extraction" }]

Rules:
- Each entry is one sentence, clinical and factual
- Max 10 entries total (merge or drop minor duplicates)
- Use realistic timestamps spread across the encounter (start ~08:12 today)
- source must be "extraction"`;

export function buildTimelinePrompt(
  entities: MedicalEntities,
  transcript: string,
  existing: TimelineEntry[]
): string {
  const today = new Date().toISOString().slice(0, 10);
  return [
    `Today: ${today}`,
    "",
    "Existing timeline:",
    JSON.stringify(existing, null, 2),
    "",
    "Entities:",
    JSON.stringify(entities, null, 2),
    "",
    "Transcript:",
    transcript.slice(-2000),
    "",
    "Return updated timeline JSON array (max 10 entries).",
  ].join("\n");
}

export function heuristicTimeline(
  entities: MedicalEntities,
  existing: TimelineEntry[]
): TimelineEntry[] {
  const entries = existing.filter((e) => e.source !== "extraction");
  const safetyEntries = existing.filter((e) => e.source === "safety");

  const baseMs = Date.now() - 20 * 60 * 1000; // ~20 min ago
  let offset = 0;

  const add = (summary: string) => {
    if (entries.some((e) => e.summary.toLowerCase().includes(summary.toLowerCase().slice(0, 30))))
      return;
    if (safetyEntries.some((e) =>
      e.summary.toLowerCase().includes(summary.toLowerCase().slice(0, 30))
    ))
      return;
    entries.push({
      id: `tl-${crypto.randomUUID().slice(0, 8)}`,
      timestamp: new Date(baseMs + offset * 90_000).toISOString(),
      summary,
      source: "extraction",
    });
    offset++;
  };

  const age = entities.demographics?.age;
  const sex = entities.demographics?.sex;

  if (age || sex) {
    const demo = [age ? `${age}yo` : null, sex ?? null].filter(Boolean).join(" ");
    add(`Patient demographics: ${demo}`);
  }

  if (entities.symptoms.some((s) => s.includes("chest pain")))
    add("Patient reports acute chest pain");
  if (entities.symptoms.some((s) => s.includes("shortness of breath")))
    add("Mild shortness of breath reported");
  if (entities.symptoms.some((s) => s.includes("left arm")))
    add("Left arm radiation noted");
  if (entities.symptoms.some((s) => s.includes("nausea")))
    add("Nausea reported");

  if (entities.conditions.some((c) => c.includes("hypertension")))
    add("Hypertension history identified");
  if (entities.conditions.some((c) => c.includes("heart valve")))
    add("Mechanical heart valve replacement history noted");
  if (entities.conditions.some((c) => c.includes("atrial fibrillation")))
    add("Atrial fibrillation history identified");

  if (entities.medications.some((m) => m.name.toLowerCase() === "lisinopril"))
    add("Lisinopril (ACE inhibitor) documented");
  if (entities.medications.some((m) => m.name.toLowerCase() === "warfarin"))
    add("Warfarin anticoagulation documented");
  if (entities.medications.some((m) => m.name.toLowerCase() === "aspirin"))
    add("Aspirin ordered/documented");

  if (entities.allergies.some((a) => a.toLowerCase().includes("penicillin")))
    add("Penicillin allergy documented");
  if (entities.allergies.length > 0 && !entities.allergies.some((a) => a.toLowerCase().includes("penicillin")))
    add(`Allergy documented: ${entities.allergies[0]}`);

  // Merge safety entries back in at the correct position
  return [...entries, ...safetyEntries]
    .sort((a, b) => a.timestamp.localeCompare(b.timestamp))
    .slice(-12);
}

File: backend/prompts/timeline.py
"""
Timeline prompts and heuristic fallback — mirrors lib/prompts/timeline.ts.
"""
from __future__ import annotations

import json
import uuid
from datetime import datetime, timedelta, timezone
from typing import List

from events import MedicalEntities, TimelineEntry


TIMELINE_SYSTEM = """You are a clinical timeline agent for pre-hospital EMS/ambulance encounters.
Produce a concise, chronological list of key clinical events from the scene.

Return ONLY a raw JSON array (no markdown):
[{ "id": string, "timestamp": string (ISO 8601), "summary": string, "source": "extraction" }]

Rules:
- Each entry is one sentence, clinical and factual
- Use ambulance context: scene arrival, patient contact, transport, en route — not ER admissions
- Max 12 entries total (merge or drop minor duplicates)
- Use realistic timestamps spread across the encounter (start from scene arrival)
- source must be "extraction\""""


def build_timeline_prompt(
    entities: MedicalEntities,
    transcript: str,
    existing: List[TimelineEntry],
) -> str:
    from events import to_dict
    today = datetime.now(timezone.utc).date().isoformat()
    return "\n".join([
        f"Today: {today}",
        "",
        "Existing timeline:",
        json.dumps([to_dict(e) for e in existing], indent=2),
        "",
        "Entities:",
        json.dumps(to_dict(entities), indent=2),
        "",
        "Transcript:",
        transcript[-2000:],
        "",
        "Return updated timeline JSON array (max 10 entries).",
    ])


def heuristic_timeline(
    entities: MedicalEntities,
    existing: List[TimelineEntry],
) -> List[TimelineEntry]:
    entries = [e for e in existing if e.source != "extraction"]
    safety_entries = [e for e in existing if e.source == "safety"]

    base_dt = datetime.now(timezone.utc) - timedelta(minutes=20)
    offset = [0]  # mutable counter

    def add(summary: str) -> None:
        low = summary.lower()[:30]
        if any(low in e.summary.lower() for e in entries):
            return
        if any(low in e.summary.lower() for e in safety_entries):
            return
        ts = (base_dt + timedelta(seconds=offset[0] * 90)).isoformat()
        entries.append(TimelineEntry(
            id=f"tl-{uuid.uuid4().hex[:8]}",
            timestamp=ts,
            summary=summary,
            source="extraction",
        ))
        offset[0] += 1

    age = entities.demographics.age if entities.demographics else None
    sex = entities.demographics.sex if entities.demographics else None

    if age or sex:
        demo = " ".join(filter(None, [f"{age}yo" if age else None, sex]))
        add(f"Patient demographics: {demo}")

    if any("chest pain" in s for s in entities.symptoms):
        add("Patient reports acute chest pain")
    if any("shortness of breath" in s for s in entities.symptoms):
        add("Mild shortness of breath reported")
    if any("left arm" in s for s in entities.symptoms):
        add("Left arm radiation noted")
    if any("nausea" in s for s in entities.symptoms):
        add("Nausea reported")

    if any("hypertension" in c for c in entities.conditions):
        add("Hypertension history identified")
    if any("heart valve" in c for c in entities.conditions):
        add("Mechanical heart valve replacement history noted")
    if any("atrial fibrillation" in c for c in entities.conditions):
        add("Atrial fibrillation history identified")

    if any(m.name.lower() == "lisinopril" for m in entities.medications):
        add("Lisinopril (ACE inhibitor) documented")
    if any(m.name.lower() == "warfarin" for m in entities.medications):
        add("Warfarin anticoagulation documented")
    if any(m.name.lower() == "aspirin" for m in entities.medications):
        add("Aspirin ordered/documented")

    if any("penicillin" in a.lower() for a in entities.allergies):
        add("Penicillin allergy documented")
    elif entities.allergies:
        add(f"Allergy documented: {entities.allergies[0]}")

    combined = sorted(entries + safety_entries, key=lambda e: e.timestamp)
    return combined[-12:]
